### DISCUSSION TITLE
Error with SetArcCostEvaluatorOfVehicle

### DISCUSSION THREAD

**Discussion by FreddyHunter** [google/or-tools#2610] [Routing (and legacy CP) questions] [answered]
**What version of OR-Tools and what language are you using?**
Version: latest
Language: Python

**Which solver are you using: Routing Solver

**What operating system and version? Linux Ubuntu 20.04

**What did you do?**
I modeled the total cost (fixed and variable) per vehicle to be included at objective function, as you can see below:
```python
    # Add Costs constraint.
    def create_cost_callback(data, vehicle_id):
        """Created call back to get total cost between locations """
#        def service_cost(from_index):
#            """ Gets the service cost for the specified location."""
#            return data["demands"][node]*data["cost_per_demand_unit"]

        def travel_cost(from_index, to_index, vehicle_id):
            """Gets the travel cost between two locations """
            # Convert from routing variable Index to time matrix NodeIndex.
            from_node = manager.IndexToNode(from_index)
            to_node = manager.IndexToNode(to_index)
            travel_cost = (data['vehicle_fix_costs'][vehicle_id] + data['distance_matrix'][from_node][to_node] * data['vehicle_var_costs'][vehicle_id])

        def cost_callback(from_index, to_index):
            """ Returns the total costs between the two nodes """
#            serv_cost = service_cost(from_node)
            trav_cost = travel_cost(from_index, to_index, vehicle_id)
            return trav_cost #+ serv_cost

        return cost_callback

    for vehicle_id in range(data['num_vehicles']):
        print(vehicle_id)
        print(int(vehicle_id))
        cost_callback = create_cost_callback(data, int(vehicle_id))
        routing.SetArcCostEvaluatorOfVehicle(cost_callback, int(vehicle_id))
```
**What did you expect to see**
During optimization I want optimizer select first the cheapest vehicles and than move forward to orders, considering the others constraints included.

**What did you see instead?**

I got the following error:
```sh
Traceback (most recent call last):
  File "optim_test9.py", line 845, in <module>
    main_optimizer()
  File "optim_test9.py", line 493, in main_optimizer
    routing.SetArcCostEvaluatorOfVehicle(cost_callback, int(vehicle_id))
  File "/usr/local/lib/python3.8/dist-packages/ortools/constraint_solver/pywrapcp.py", line 5212, in SetArcCostEvaluatorOfVehicle
    return _pywrapcp.RoutingModel_SetArcCostEvaluatorOfVehicle(self, evaluator_index, vehicle)
TypeError: in method 'RoutingModel_SetArcCostEvaluatorOfVehicle', argument 2 of type 'int'
```

Make sure you include information that can help us debug (full error message, model Proto).

**Anything else we should know about your project / environment**

**Comment by Mizux** [comment, marked as answer]
you didn't "register"  your callback, you are passing a pointer to the function instead of an id (aka evaluator_idx)
To get it you need to call `cost_callback_index = routing.RegisterTransitCallback(cost_callback)`
e.g.
https://github.com/google/or-tools/blob/b37d9c786b69128f3505f15beca09e89bf078a89/ortools/constraint_solver/samples/simple_routing_program.py#L45-L57

**Reply by FreddyHunter** [reply]
Hi Mizux, 

thanks for your reply. I adjusted the code and now it is running without reporting error, but it seems not be taking in considerations the cost per vehicle. I mean, I was expecting to be selected the cheapest vehicles first than most expensive ones, but as you can see at the print solution that was not the case. The model is like that:

 data['num_vehicles'] = 7
data['vehicle_capacities'] = [15, 15, 15, 15, 15, 15, 15]
 data['vehicle_fix_costs'] = [50, 50, 50, 50, 50, 100, 100]
 data['vehicle_var_costs'] = [2, 2, 1, 1, 1, 5, 5]
data['vehicle_max_costs'] = 100

    # Add Distance constraint.
    def distance_callback(from_index, to_index):
        """Returns the distance between the two nodes."""
        # Convert from routing variable Index to time matrix NodeIndex.
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return int(data['distance_matrix'][from_node][to_node]/1000)

    transit_callback_index = routing.RegisterTransitCallback(distance_callback)

    # Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

#    distance_callback_index = routing.RegisterTransitCallback(distance_callback)
    dimension_name = 'Distance'
    vehicle_max_distance = data['vehicle_max_distances']
    routing.AddDimensionWithVehicleCapacity(
        transit_callback_index,
        0,  # null capacity slack
        vehicle_max_distance,  # vehicle maximum capacities
        True,  # start cumul to zero
        dimension_name)
    distance_dimension = routing.GetDimensionOrDie(dimension_name)

    # Add Costs constraint.
    def create_cost_callback(data, vehicle_id):
        """Created call back to get total cost between locations """
        def service_cost(from_index):
            """ Gets the service cost for the specified location."""
#            return data["demands"][node]*data["cost_per_demand_unit"]
            from_node = manager.IndexToNode(from_index)
            to_node = manager.IndexToNode(to_index)
            if data['time_per_demand_parameter'] == 0:
               (data['time_per_demand'][from_node] * data['cost_per_demand'][from_node])
            else:
               (((data['demands'][from_node][0] + data['demands'][from_node][1] + data['demands'][from_node][2])* data['time_per_demand'][from_node]) * data['cost_per_demand'][from_node])

        def travel_cost(from_index, to_index, vehicle_id):
            """Gets the travel cost between two locations """
            # Convert from routing variable Index to time matrix NodeIndex.
            from_node = manager.IndexToNode(from_index)
            to_node = manager.IndexToNode(to_index)
            travel_cost = (data['vehicle_fix_costs'][vehicle_id] + data['distance_matrix'][from_node][to_node] * data['vehicle_var_costs'][vehicle_id])


        def cost_callback(from_index, to_index):
            """ Returns the total costs between the two nodes """
            serv_cost = service_cost(from_index)
            trav_cost = travel_cost(from_index, to_index, vehicle_id)
            return int(trav_cost + serv_cost)

        return cost_callback

    cost_callback_index = routing.RegisterTransitCallback(create_cost_callback)

    for vehicle_id in range(data['num_vehicles']):
        routing.SetArcCostEvaluatorOfVehicle(cost_callback_index, int(vehicle_id))

    # Add Vehicle Costs constraint.
    dimension_name = 'Cost'
    vehicle_max_cost = data['vehicle_max_costs']
    routing.AddDimension(
        cost_callback_index,
        0,  # null capacity slack
        vehicle_max_cost,  # vehicle maximum cost
        True,  # start cumul to zero
        dimension_name)
    cost_dimension = routing.GetDimensionOrDie(dimension_name)

Solver status:  1
Objective: 0
Dropped nodes:
Route for vehicle 0:
0 Time(0,0) -> 0 Time(0,0)
Time of the route: 0min

Load for Vehicle 0:
 0 Load (0) ->  0 Load (0)
Load of the route: 0

Distance for Vehicle 0:
 0 Distance(0) ->  0 Distance(0)
Distance of the route: 0m

Cost for Vehicle 0:
 0 Cost(0) ->  0 Cost(0)
Cost of the route: $50

Route for vehicle 1:
0 Time(0,0) -> 0 Time(0,0)
Time of the route: 0min

Load for Vehicle 1:
 0 Load (0) ->  0 Load (0)
Load of the route: 0

Distance for Vehicle 1:
 0 Distance(0) ->  0 Distance(0)
Distance of the route: 0m

Cost for Vehicle 1:
 0 Cost(0) ->  0 Cost(0)
Cost of the route: $50

Route for vehicle 2:
0 Time(0,0) -> 0 Time(0,0)
Time of the route: 0min

Load for Vehicle 2:
 0 Load (0) ->  0 Load (0)
Load of the route: 0

Distance for Vehicle 2:
 0 Distance(0) ->  0 Distance(0)
Distance of the route: 0m

Cost for Vehicle 2:
 0 Cost(0) ->  0 Cost(0)
Cost of the route: $50

Route for vehicle 3:
0 Time(360,360) -> 4 Time(479,480) -> 3 Time(478,480) -> 2 Time(479,480) -> 1 Time(478,480) -> 0 Time(478,478)
Time of the route: 478min

Load for Vehicle 3:
 0 Load (0) ->  4 Load (4) ->  3 Load (7) ->  2 Load (9) ->  1 Load (10) ->  0 Load (10)
Load of the route: 10

Distance for Vehicle 3:
 0 Distance(4) ->  4 Distance(5) ->  3 Distance(6) ->  2 Distance(7) ->  1 Distance(8) ->  0 Distance(8)
Distance of the route: 8m

Cost for Vehicle 3:
 0 Cost(4) ->  4 Cost(5) ->  3 Cost(6) ->  2 Cost(7) ->  1 Cost(8) ->  0 Cost(8)
Cost of the route: $58

Route for vehicle 4:
0 Time(360,360) -> 8 Time(479,480) -> 7 Time(478,480) -> 0 Time(478,478)
Time of the route: 478min

Load for Vehicle 4:
 0 Load (0) ->  8 Load (8) ->  7 Load (15) ->  0 Load (15)
Load of the route: 15

Distance for Vehicle 4:
 0 Distance(8) ->  8 Distance(9) ->  7 Distance(16) ->  0 Distance(16)
Distance of the route: 16m

Cost for Vehicle 4:
 0 Cost(8) ->  8 Cost(9) ->  7 Cost(16) ->  0 Cost(16)
Cost of the route: $66

Route for vehicle 5:
0 Time(360,360) -> 9 Time(478,480) -> 6 Time(479,480) -> 0 Time(479,479)
Time of the route: 479min

Load for Vehicle 5:
 0 Load (0) ->  9 Load (9) ->  6 Load (15) ->  0 Load (15)
Load of the route: 15

Distance for Vehicle 5:
 0 Distance(9) ->  9 Distance(12) ->  6 Distance(18) ->  0 Distance(18)
Distance of the route: 18m

Cost for Vehicle 5:
 0 Cost(45) ->  9 Cost(60) ->  6 Cost(90) ->  0 Cost(90)
Cost of the route: $190

Route for vehicle 6:
0 Time(360,360) -> 10 Time(479,480) -> 5 Time(478,480) -> 0 Time(478,478)
Time of the route: 478min

Load for Vehicle 6:
 0 Load (0) ->  10 Load (10) ->  5 Load (15) ->  0 Load (15)
Load of the route: 15

Distance for Vehicle 6:
 0 Distance(10) ->  10 Distance(15) ->  5 Distance(20) ->  0 Distance(20)
Distance of the route: 20m

Cost for Vehicle 6:
 0 Cost(50) ->  10 Cost(75) ->  5 Cost(100) ->  0 Cost(100)
Cost of the route: $200

Total vehicles used of all routes: 4 vehicles
Total time of all routes: 1913min
Total Load of all routes: 55
Total Distance of all routes: 62m
Total Cost of all routes: $214

**Comment by FreddyHunter** [comment]
Just complementing... It seems to be conflicting with something. As you can notice at the solution printed before, objetive is returning 0. If I comment "Add Cost constraints" and " Add Vehicle Cost constraints" sessions on the code the objetive function returns different than zero.

Below is the begining of the code:

    manager = pywrapcp.RoutingIndexManager(len(data['distance_matrix']),
                                           data['num_vehicles'], data['starts'], data['ends'])

    routing = pywrapcp.RoutingModel(manager)

    # Create and register a transit callback.
    # Add Distance constraint.
    def distance_callback(from_index, to_index):
        """Returns the distance between the two nodes."""
        # Convert from routing variable Index to time matrix NodeIndex.
        from_node = manager.IndexToNode(from_index)
        to_node = manager.IndexToNode(to_index)
        return int(data['distance_matrix'][from_node][to_node]/1000)

    transit_callback_index = routing.RegisterTransitCallback(distance_callback)

    # Define cost of each arc.
    routing.SetArcCostEvaluatorOfAllVehicles(transit_callback_index)

#    distance_callback_index = routing.RegisterTransitCallback(distance_callback)
    dimension_name = 'Distance'
     vehicle_max_distance = data['vehicle_max_distances']
    routing.AddDimensionWithVehicleCapacity(
        transit_callback_index,
        0,  # null capacity slack
        vehicle_max_distance,  # vehicle maximum capacities
        True,  # start cumul to zero
        dimension_name)
    distance_dimension = routing.GetDimensionOrDie(dimension_name)
